// @emnapi/runtime/dist/emnapi.iife.js
var emnapi = (function (exports) {

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise, SuppressedError, Symbol, Iterator */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
        var e = new Error(message);
        return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
    };

    var externalValue = new WeakMap();
    /** @public */
    function isExternal(object) {
        return externalValue.has(object);
    }
    /** @public */ // eslint-disable-next-line @typescript-eslint/no-redeclare
    var External = (function () {
        function External(value) {
            Object.setPrototypeOf(this, null);
            externalValue.set(this, value);
        }
        External.prototype = null;
        return External;
    })();
    /** @public */
    function getExternalValue(external) {
        if (!isExternal(external)) {
            throw new TypeError('not external');
        }
        return externalValue.get(external);
    }

    var supportNewFunction = /*#__PURE__*/ (function () {
        var f;
        try {
            f = new Function();
        }
        catch (_) {
            return false;
        }
        return typeof f === 'function';
    })();
    var _global = /*#__PURE__*/ (function () {
        if (typeof globalThis !== 'undefined')
            return globalThis;
        var g = (function () { return this; })();
        if (!g && supportNewFunction) {
            try {
                g = new Function('return this')();
            }
            catch (_) { }
        }
        if (!g) {
            if (typeof __webpack_public_path__ === 'undefined') {
                if (typeof global !== 'undefined')
                    return global;
            }
            if (typeof window !== 'undefined')
                return window;
            if (typeof self !== 'undefined')
                return self;
        }
        return g;
    })();
    var TryCatch = /*#__PURE__*/ (function () {
        function TryCatch() {
            this._exception = undefined;
            this._caught = false;
        }
        TryCatch.prototype.isEmpty = function () {
            return !this._caught;
        };
        TryCatch.prototype.hasCaught = function () {
            return this._caught;
        };
        TryCatch.prototype.exception = function () {
            return this._exception;
        };
        TryCatch.prototype.setError = function (err) {
            this._caught = true;
            this._exception = err;
        };
        TryCatch.prototype.reset = function () {
            this._caught = false;
            this._exception = undefined;
        };
        TryCatch.prototype.extractException = function () {
            var e = this._exception;
            this.reset();
            return e;
        };
        return TryCatch;
    }());
    var canSetFunctionName = /*#__PURE__*/ (function () {
        var _a;
        try {
            return Boolean((_a = Object.getOwnPropertyDescriptor(Function.prototype, 'name')) === null || _a === void 0 ? void 0 : _a.configurable);
        }
        catch (_) {
            return false;
        }
    })();
    var supportReflect = typeof Reflect === 'object';
    var supportFinalizer = (typeof FinalizationRegistry !== 'undefined') && (typeof WeakRef !== 'undefined');
    var supportWeakSymbol = /*#__PURE__*/ (function () {
        try {
            // eslint-disable-next-line symbol-description
            var sym = Symbol();
            // eslint-disable-next-line no-new
            new WeakRef(sym);
            new WeakMap().set(sym, undefined);
        }
        catch (_) {
            return false;
        }
        return true;
    })();
    var supportBigInt = typeof BigInt !== 'undefined';
    function isReferenceType(v) {
        return (typeof v === 'object' && v !== null) || typeof v === 'function';
    }
    var _require = /*#__PURE__*/ (function () {
        var nativeRequire;
        if (typeof __webpack_public_path__ !== 'undefined') {
            nativeRequire = (function () {
                return typeof __non_webpack_require__ !== 'undefined' ? __non_webpack_require__ : undefined;
            })();
        }
        else {
            nativeRequire = (function () {
                return typeof __webpack_public_path__ !== 'undefined' ? (typeof __non_webpack_require__ !== 'undefined' ? __non_webpack_require__ : undefined) : (typeof require !== 'undefined' ? require : undefined);
            })();
        }
        return nativeRequire;
    })();
    var _MessageChannel = typeof MessageChannel === 'function'
        ? MessageChannel
        : /*#__PURE__*/ (function () {
            try {
                return _require('worker_threads').MessageChannel;
            }
            catch (_) { }
            return undefined;
        })();
    var _setImmediate = typeof setImmediate === 'function'
        ? setImmediate
        : function (callback) {
            if (typeof callback !== 'function') {
                throw new TypeError('The "callback" argument must be of type function');
            }
            if (_MessageChannel) {
                var channel_1 = new _MessageChannel();
                channel_1.port1.onmessage = function () {
                    channel_1.port1.onmessage = null;
                    channel_1 = undefined;
                    callback();
                };
                channel_1.port2.postMessage(null);
            }
            else {
                setTimeout(callback, 0);
            }
        };
    var _Buffer = typeof Buffer === 'function'
        ? Buffer
        : /*#__PURE__*/ (function () {
            try {
                return _require('buffer').Buffer;
            }
            catch (_) { }
            return undefined;
        })();
    var version = "1.4.0";
    var NODE_API_SUPPORTED_VERSION_MIN = 1 /* Version.NODE_API_SUPPORTED_VERSION_MIN */;
    var NODE_API_SUPPORTED_VERSION_MAX = 10 /* Version.NODE_API_SUPPORTED_VERSION_MAX */;
    var NAPI_VERSION_EXPERIMENTAL = 2147483647 /* Version.NAPI_VERSION_EXPERIMENTAL */;
    var NODE_API_DEFAULT_MODULE_API_VERSION = 8 /* Version.NODE_API_DEFAULT_MODULE_API_VERSION */;

    var Handle = /*#__PURE__*/ (function () {
        function Handle(id, value) {
            this.id = id;
            this.value = value;
        }
        Handle.prototype.data = function () {
            return getExternalValue(this.value);
        };
        Handle.prototype.isNumber = function () {
            return typeof this.value === 'number';
        };
        Handle.prototype.isBigInt = function () {
            return typeof this.value === 'bigint';
        };
        Handle.prototype.isString = function () {
            return typeof this.value === 'string';
        };
        Handle.prototype.isFunction = function () {
            return typeof this.value === 'function';
        };
        Handle.prototype.isExternal = function () {
            return isExternal(this.value);
        };
        Handle.prototype.isObject = function () {
            return typeof this.value === 'object' && this.value !== null;
        };
        Handle.prototype.isArray = function () {
            return Array.isArray(this.value);
        };
        Handle.prototype.isArrayBuffer = function () {
            return (this.value instanceof ArrayBuffer);
        };
        Handle.prototype.isTypedArray = function () {
            return (ArrayBuffer.isView(this.value)) && !(this.value instanceof DataView);
        };
        Handle.prototype.isBuffer = function (BufferConstructor) {
            if (ArrayBuffer.isView(this.value))
                return true;
            BufferConstructor !== null && BufferConstructor !== void 0 ? BufferConstructor : (BufferConstructor = _Buffer);
            return typeof BufferConstructor === 'function' && BufferConstructor.isBuffer(this.value);
        };
        Handle.prototype.isDataView = function () {
            return (this.value instanceof DataView);
        };
        Handle.prototype.isDate = function () {
            return (this.value instanceof Date);
        };
        Handle.prototype.isPromise = function () {
            return (this.value instanceof Promise);
        };
        Handle.prototype.isBoolean = function () {
            return typeof this.value === 'boolean';
        };
        Handle.prototype.isUndefined = function () {
            return this.value === undefined;
        };
        Handle.prototype.isSymbol = function () {
            return typeof this.value === 'symbol';
        };
        Handle.prototype.isNull = function () {
            return this.value === null;
        };
        Handle.prototype.dispose = function () {
            this.value = undefined;
        };
        return Handle;
    }());
    var ConstHandle = /*#__PURE__*/ (function (_super) {
        __extends(ConstHandle, _super);
        function ConstHandle(id, value) {
            return _super.call(this, id, value) || this;
        }
        ConstHandle.prototype.dispose = function () { };
        return ConstHandle;
    }(Handle));
    var HandleStore = /*#__PURE__*/ (function () {
        function HandleStore() {
            this._values = [
                undefined,
                HandleStore.UNDEFINED,
                HandleStore.NULL,
                HandleStore.FALSE,
                HandleStore.TRUE,
                HandleStore.GLOBAL
            ];
            this._next = HandleStore.MIN_ID;
        }
        HandleStore.prototype.push = function (value) {
            var h;
            var next = this._next;
            var values = this._values;
            if (next < values.length) {
                h = values[next];
                h.value = value;
            }
            else {
                h = new Handle(next, value);
                values[next] = h;
            }
            this._next++;
            return h;
        };
        HandleStore.prototype.erase = function (start, end) {
            this._next = start;
            var values = this._values;
            for (var i = start; i < end; ++i) {
                values[i].dispose();
            }
        };
        HandleStore.prototype.get = function (id) {
            return this._values[id];
        };
        HandleStore.prototype.swap = function (a, b) {
            var values = this._values;
            var h = values[a];
            values[a] = values[b];
            values[a].id = Number(a);
            values[b] = h;
            h.id = Number(b);
        };
        HandleStore.prototype.dispose = function () {
            this._values.length = HandleStore.MIN_ID;
            this._next = HandleStore.MIN_ID;
        };
        HandleStore.UNDEFINED = new ConstHandle(1 /* GlobalHandle.UNDEFINED */, undefined);
        HandleStore.NULL = new ConstHandle(2 /* GlobalHandle.NULL */, null);
        HandleStore.FALSE = new ConstHandle(3 /* GlobalHandle.FALSE */, false);
        HandleStore.TRUE = new ConstHandle(4 /* GlobalHandle.TRUE */, true);
        HandleStore.GLOBAL = new ConstHandle(5 /* GlobalHandle.GLOBAL */, _global);
        HandleStore.MIN_ID = 6;
        return HandleStore;
    }());

    var HandleScope = /*#__PURE__*/ (function () {
        function HandleScope(handleStore, id, parentScope, start, end) {
            if (end === void 0) { end = start; }
            this.handleStore = handleStore;
            this.id = id;
            this.parent = parentScope;
            this.child = null;
            if (parentScope !== null)
                parentScope.child = this;
            this.start = start;
            this.end = end;
            this._escapeCalled = false;
            this.callbackInfo = {
                thiz: undefined,
                data: 0,
                args: undefined,
                fn: undefined
            };
        }
        HandleScope.prototype.add = function (value) {
            var h = this.handleStore.push(value);
            this.end++;
            return h;
        };
        HandleScope.prototype.addExternal = function (data) {
            return this.add(new External(data));
        };
        HandleScope.prototype.dispose = function () {
            if (this._escapeCalled)
                this._escapeCalled = false;
            if (this.start === this.end)
                return;
            this.handleStore.erase(this.start, this.end);
        };
        HandleScope.prototype.escape = function (handle) {
            if (this._escapeCalled)
                return null;
            this._escapeCalled = true;
            if (handle < this.start || handle >= this.end) {
                return null;
            }
            this.handleStore.swap(handle, this.start);
            var h = this.handleStore.get(this.start);
            this.start++;
            this.parent.end++;
            return h;
        };
        HandleScope.prototype.escapeCalled = function () {
            return this._escapeCalled;
        };
        return HandleScope;
    }());

    var ScopeStore = /*#__PURE__*/ (function () {
        function ScopeStore() {
            this._rootScope = new HandleScope(null, 0, null, 1, HandleStore.MIN_ID);
            this.currentScope = this._rootScope;
            this._values = [undefined];
        }
        ScopeStore.prototype.get = function (id) {
            return this._values[id];
        };
        ScopeStore.prototype.openScope = function (envObject) {
            var currentScope = this.currentScope;
            var scope = currentScope.child;
            if (scope !== null) {
                scope.start = scope.end = currentScope.end;
            }
            else {
                var id = currentScope.id + 1;
                scope = new HandleScope(envObject.ctx.handleStore, id, currentScope, currentScope.end);
                this._values[id] = scope;
            }
            this.currentScope = scope;
            envObject.openHandleScopes++;
            return scope;
        };
        ScopeStore.prototype.closeScope = function (envObject) {
            if (envObject.openHandleScopes === 0)
                return;
            var scope = this.currentScope;
            this.currentScope = scope.parent;
            scope.dispose();
            envObject.openHandleScopes--;
        };
        ScopeStore.prototype.dispose = function () {
            this.currentScope = this._rootScope;
            this._values.length = 1;
        };
        return ScopeStore;
    }());

    var RefTracker = /*#__PURE__*/ (function () {
        function RefTracker() {
            this._next = null;
            this._prev = null;
        }
        /** @virtual */
        RefTracker.prototype.dispose = function () { };
        /** @virtual */
        RefTracker.prototype.finalize = function () { };
        RefTracker.prototype.link = function (list) {
            this._prev = list;
            this._next = list._next;
            if (this._next !== null) {
                this._next._prev = this;
            }
            list._next = this;
        };
        RefTracker.prototype.unlink = function () {
            if (this._prev !== null) {
                this._prev._next = this._next;
            }
            if (this._next !== null) {
                this._next._prev = this._prev;
            }
            this._prev = null;
            this._next = null;
        };
        RefTracker.finalizeAll = function (list) {
            while (list._next !== null) {
                list._next.finalize();
            }
        };
        return RefTracker;
    }());

    var Finalizer = /*#__PURE__*/ (function () {
        function Finalizer(envObject, _finalizeCallback, _finalizeData, _finalizeHint) {
            if (_finalizeCallback === void 0) { _finalizeCallback = 0; }
            if (_finalizeData === void 0) { _finalizeData = 0; }
            if (_finalizeHint === void 0) { _finalizeHint = 0; }
            this.envObject = envObject;
            this._finalizeCallback = _finalizeCallback;
            this._finalizeData = _finalizeData;
            this._finalizeHint = _finalizeHint;
            this._makeDynCall_vppp = envObject.makeDynCall_vppp;
        }
        Finalizer.prototype.callback = function () { return this._finalizeCallback; };
        Finalizer.prototype.data = function () { return this._finalizeData; };
        Finalizer.prototype.hint = function () { return this._finalizeHint; };
        Finalizer.prototype.resetEnv = function () {
            this.envObject = undefined;
        };
        Finalizer.prototype.resetFinalizer = function () {
            this._finalizeCallback = 0;
            this._finalizeData = 0;
            this._finalizeHint = 0;
        };
        Finalizer.prototype.callFinalizer = function () {
            var finalize_callback = this._finalizeCallback;
            var finalize_data = this._finalizeData;
            var finalize_hint = this._finalizeHint;
            this.resetFinalizer();
            if (!finalize_callback)
                return;
            var fini = Number(finalize_callback);
            if (!this.envObject) {
                this._makeDynCall_vppp(fini)(0, finalize_data, finalize_hint);
            }
            else {
                this.envObject.callFinalizer(fini, finalize_data, finalize_hint);
            }
        };
        Finalizer.prototype.dispose = function () {
            this.envObject = undefined;
            this._makeDynCall_vppp = undefined;
        };
        return Finalizer;
    }());

    var TrackedFinalizer = /*#__PURE__*/ (function (_super) {
        __extends(TrackedFinalizer, _super);
        function TrackedFinalizer(envObject, finalize_callback, finalize_data, finalize_hint) {
            var _this = _super.call(this) || this;
            _this._finalizer = new Finalizer(envObject, finalize_callback, finalize_data, finalize_hint);
            return _this;
        }
        TrackedFinalizer.create = function (envObject, finalize_callback, finalize_data, finalize_hint) {
            var finalizer = new TrackedFinalizer(envObject, finalize_callback, finalize_data, finalize_hint);
            finalizer.link(envObject.finalizing_reflist);
            return finalizer;
        };
        TrackedFinalizer.prototype.data = function () {
            return this._finalizer.data();
        };
        TrackedFinalizer.prototype.dispose = function () {
            if (!this._finalizer)
                return;
            this.unlink();
            this._finalizer.envObject.dequeueFinalizer(this);
            this._finalizer.dispose();
            this._finalizer = undefined;
            _super.prototype.dispose.call(this);
        };
        TrackedFinalizer.prototype.finalize = function () {
            this.unlink();
            var error;
            var caught = false;
            try {
                this._finalizer.callFinalizer();
            }
            catch (err) {
                caught = true;
                error = err;
            }
            this.dispose();
            if (caught) {
                throw error;
            }
        };
        return TrackedFinalizer;
    }(RefTracker));

    function throwNodeApiVersionError(moduleName, moduleApiVersion) {
        var errorMessage = "".concat(moduleName, " requires Node-API version ").concat(moduleApiVersion, ", but this version of Node.js only supports version ").concat(NODE_API_SUPPORTED_VERSION_MAX, " add-ons.");
        throw new Error(errorMessage);
    }
    function handleThrow(envObject, value) {
        if (envObject.terminatedOrTerminating()) {
            return;
        }
        throw value;
    }
    var Env = /*#__PURE__*/ (function () {
        function Env(ctx, moduleApiVersion, makeDynCall_vppp, makeDynCall_vp, abort) {
            this.ctx = ctx;
            this.moduleApiVersion = moduleApiVersion;
            this.makeDynCall_vppp = makeDynCall_vppp;
            this.makeDynCall_vp = makeDynCall_vp;
            this.abort = abort;
            this.openHandleScopes = 0;
            this.instanceData = null;
            this.tryCatch = new TryCatch();
            this.refs = 1;
            this.reflist = new RefTracker();
            this.finalizing_reflist = new RefTracker();
            this.pendingFinalizers = [];
            this.lastError = {
                errorCode: 0 /* napi_status.napi_ok */,
                engineErrorCode: 0,
                engineReserved: 0
            };
            this.inGcFinalizer = false;
            this._bindingMap = new WeakMap();
            this.id = 0;
        }
        /** @virtual */
        Env.prototype.canCallIntoJs = function () {
            return true;
        };
        Env.prototype.terminatedOrTerminating = function () {
            return !this.canCallIntoJs();
        };
        Env.prototype.ref = function () {
            this.refs++;
        };
        Env.prototype.unref = function () {
            this.refs--;
            if (this.refs === 0) {
                this.dispose();
            }
        };
        Env.prototype.ensureHandle = function (value) {
            return this.ctx.ensureHandle(value);
        };
        Env.prototype.ensureHandleId = function (value) {
            return this.ensureHandle(value).id;
        };
        Env.prototype.clearLastError = function () {
            var lastError = this.lastError;
            if (lastError.errorCode !== 0 /* napi_status.napi_ok */)
                lastError.errorCode = 0 /* napi_status.napi_ok */;
            if (lastError.engineErrorCode !== 0)
                lastError.engineErrorCode = 0;
            if (lastError.engineReserved !== 0)
                lastError.engineReserved = 0;
            return 0 /* napi_status.napi_ok */;
        };
        Env.prototype.setLastError = function (error_code, engine_error_code, engine_reserved) {
            if (engine_error_code === void 0) { engine_error_code = 0; }
            if (engine_reserved === void 0) { engine_reserved = 0; }
            var lastError = this.lastError;
            if (lastError.errorCode !== error_code)
                lastError.errorCode = error_code;
            if (lastError.engineErrorCode !== engine_error_code)
                lastError.engineErrorCode = engine_error_code;
            if (lastError.engineReserved !== engine_reserved)
                lastError.engineReserved = engine_reserved;
            return error_code;
        };
        Env.prototype.getReturnStatus = function () {
            return !this.tryCatch.hasCaught() ? 0 /* napi_status.napi_ok */ : this.setLastError(10 /* napi_status.napi_pending_exception */);
        };
        Env.prototype.callIntoModule = function (fn, handleException) {
            if (handleException === void 0) { handleException = handleThrow; }
            var openHandleScopesBefore = this.openHandleScopes;
            this.clearLastError();
            var r = fn(this);
            if (openHandleScopesBefore !== this.openHandleScopes) {
                this.abort('open_handle_scopes != open_handle_scopes_before');
            }
            if (this.tryCatch.hasCaught()) {
                var err = this.tryCatch.extractException();
                handleException(this, err);
            }
            return r;
        };
        Env.prototype.invokeFinalizerFromGC = function (finalizer) {
            if (this.moduleApiVersion !== NAPI_VERSION_EXPERIMENTAL) {
                this.enqueueFinalizer(finalizer);
            }
            else {
                var saved = this.inGcFinalizer;
                this.inGcFinalizer = true;
                try {
                    finalize
